link/trace_event.go

```
@@ -0,0 +1,322 @@
+type TraceEvent struct {
+	// Group and name of the tracepoint/kprobe/uprobe.
+	group string
+	name  string
+
+	// PMU event ID read from sysfs. Valid IDs are non-zero.
+	pmuID uint64
+	// ID of the trace event read from tracefs. Valid IDs are non-zero.
+	tracefsID uint64
+
+	// True for kretprobes/uretprobes.
+	ret bool
+
+	// Type of program this trace event will accept to be attached.
+	progType ebpf.ProgramType
+}
+
+// Close releases the resources associated with the TraceEvent.
+func (te *TraceEvent) Close() error {
+	if te.group == "" || te.name == "" {
+		return errors.New("cannot close an empty trace event")
+	}
+
+	switch t := te.progType; t {
+	case ebpf.Kprobe:
+		return te.closeKprobe()
+	case ebpf.TracePoint:
+		return nil
+	default:
+		return fmt.Errorf("unknown program type for trace event: %s", t)
+	}
+}
+
+// Attach opens a new perf event for the TraceEvent and attaches the given
+// program to it.
+func (te *TraceEvent) Attach(program *ebpf.Program) (Link, error) {
+	if program == nil {
+		return nil, errors.New("cannot attach a nil program")
+	}
+	if t := program.Type(); t != te.progType {
+		return nil, fmt.Errorf("invalid program type (expected %s): %s", te.progType, t)
+	}
+	if program.FD() < 0 {
+		return nil, fmt.Errorf("invalid program: %w", internal.ErrClosedFd)
+	}
+
+	var ifd *internal.FD
+
+	switch {
+	case te.pmuID != 0:
+		// Open a PMU perf event based on a sysfs PMU type.
+		fd, err := te.perfEventOpenPMU()
+		if err != nil {
+			return nil, fmt.Errorf("open PMU trace event: %w", err)
+		}
+		ifd = fd
+
+	case te.tracefsID != 0:
+		// Open a perf event based on a tracefs trace event ID.
+		fd, err := te.perfEventOpenTraceFS()
+		if err != nil {
+			return nil, fmt.Errorf("open tracefs perf event: %w", err)
+		}
+		ifd = fd
+
+	default:
+		return nil, errors.New("need a PMU type or tracefs ID")
+	}
+
+	kfd, err := ifd.Value()
+	if err != nil {
+		return nil, fmt.Errorf("getting perf event fd: %w", err)
```

This and the following return leak `ifd`.

Nit: I think it'd be ok to ignore the error here and just use kfd.

---

Good catch. fixed.